modules/upload/upload.go

```diff
@@ -0,0 +1,97 @@
+// Copyright 2019 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package upload
+
+import (
+	"net/http"
+	"path"
+	"regexp"
+	"strings"
+
+	"code.gitea.io/gitea/modules/context"
+	"code.gitea.io/gitea/modules/log"
+	"code.gitea.io/gitea/modules/secret"
+	"code.gitea.io/gitea/modules/setting"
+)
+
+// ErrFileTypeForbidden not allowed file type error
+type ErrFileTypeForbidden struct {
+	Type string
+}
+
+// IsErrFileTypeForbidden checks if an error is a ErrFileTypeForbidden.
+func IsErrFileTypeForbidden(err error) bool {
+	_, ok := err.(ErrFileTypeForbidden)
+	return ok
+}
+
+func (err ErrFileTypeForbidden) Error() string {
+	return "This file extension or type is not allowed to be uploaded."
+}
+
+var mimeTypeSuffixRe = regexp.MustCompile(`;.*$`)
+var wildcardTypeRe = regexp.MustCompile(`^[a-z]+/\*$`)
+
+// Verify validates whether a file is allowed to be uploaded.
+func Verify(buf []byte, fileName string, allowedTypesStr string) error {
+	allowedTypesStr = strings.ReplaceAll(allowedTypesStr, "|", ",") // compat for old config format
+
+	allowedTypes := []string{}
+	for _, entry := range strings.Split(allowedTypesStr, ",") {
+		entry = strings.ToLower(strings.TrimSpace(entry))
+		if entry != "" {
+			allowedTypes = append(allowedTypes, entry)
+		}
+	}
+
+	if len(allowedTypes) == 0 {
+		return nil // everything is allowed
+	}
+
+	fullMimeType := http.DetectContentType(buf)
+	mimeType := strings.TrimSpace(mimeTypeSuffixRe.ReplaceAllString(fullMimeType, ""))
+	extension := strings.ToLower(path.Ext(fileName))
+
+	// https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/file#Unique_file_type_specifiers
+	for _, allowEntry := range allowedTypes {
+		if allowEntry == "*/*" {
+			return nil // everything allowed
+		} else if strings.HasPrefix(allowEntry, ".") && allowEntry == extension {
+			return nil // extension is allowed
+		} else if mimeType == allowEntry {
+			return nil // mime type is allowed
+		} else if wildcardTypeRe.MatchString(allowEntry) && strings.HasPrefix(mimeType, allowEntry[:len(allowEntry)-1]) {
+			return nil // wildcard match, e.g. image/*
+		}
+	}
+
+	log.Info("Attachment with type %s blocked from upload", fullMimeType)
+	return ErrFileTypeForbidden{Type: fullMimeType}
+}
+
+// RenderUploadSettings renders template values for dropzone
+func RenderUploadSettings(ctx *context.Context, uploadType string) {
+	if uploadType == "release" {
+		secret, _ := secret.EncryptSecret(setting.SecretKey, uploadType)
```

Comment:
what is the point of this secret?

Reply:
A weak protection to prevent users of issue attachments to circumvent the file type restrictions (e.g. by default we allow a bunch of extension on issue attachments and everything on release attachments). The root of the issue is that both POSTs go to the same shared `/attachments` endpoint and a user could just post a secret obtained from the release page to upload all types to an issue. Idea was to encrypt the repo id into the secret but that would require additional state to be held on the server for verification. I'm generally not happy with this solution, maybe there's something I'm missing (besides the idea of making the release/comment POSTs send only 1 POST instead of currently two).

Reply:
but it will never change so there is not really use for that. Encryption is not good for such protection anyway. I would propose to use `md5(secret+csrf+'<type>')` for secret. Later on you can verify by hashing secred + received csrf + all types to see which hash does match received secret

Reply:
Thing is I can't really generate a secret without storing it somewhere (probably in cache) for later verification when the /attachment POST comes in and that storage would need to be written every time someone opens a issue/pull request because we always render the upload form on those pages so cache would get filled with useless keys that will need to be eviced later on. I would prefer a stateless mechanism (or probably just combine the two POSTS but that will be more work).

Reply:
So I think what I'll do is.

1. Create endpoint `POST /user/repo/attachment?type=comment`
2. Do permission check and file type checks when that POST is received with the repo in context
3. If both checks pass, create the attachment

I guess for types "repo" and "release" we can check for write permission to the repo and for "comment" just a check if repo is not archived.

That way, we don't even need to encrypt anything.

Reply:
I've removed the encryption and split out the POST /attachments into separate ones for  issue/release.

Reply:
You would be able verify as secret is in settings (encryption secret previously) and csrf is also sent on posts but having different urls is probably better